pkg/controllers/walruscore/connector_binding.go

```diff
@@ -0,0 +1,90 @@
+package walruscore
+
+import (
+	"context"
+	"strings"
+
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlcli "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlevent "sigs.k8s.io/controller-runtime/pkg/event"
+	ctrlhandler "sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	ctrlreconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	walruscore "github.com/seal-io/walrus/pkg/apis/walruscore/v1"
+	"github.com/seal-io/walrus/pkg/controller"
+)
+
+// ConnectorBindingReconciler reconciles a v1.ConnectorBinding object.
+type ConnectorBindingReconciler struct {
+	client ctrlcli.Client
+}
+
+var _ ctrlreconcile.Reconciler = (*ConnectorBindingReconciler)(nil)
+
+func (r *ConnectorBindingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := new(walruscore.ConnectorBinding)
+
+	err := r.client.Get(ctx, req.NamespacedName, obj)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	conn := &walruscore.Connector{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      obj.Spec.Connector.Name,
+			Namespace: obj.Spec.Connector.Namespace,
+		},
+	}
+
+	err = r.client.Get(ctx, ctrlcli.ObjectKeyFromObject(conn), conn)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	obj.Status.Type = conn.Spec.Type
+	obj.Status.Category = conn.Spec.Category
+
+	err = r.client.Status().Update(ctx, obj)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ConnectorBindingReconciler) SetupController(_ context.Context, opts controller.SetupOptions) error {
+	r.client = opts.Manager.GetClient()
+
+	return ctrl.NewControllerManagedBy(opts.Manager).
+		For(&walruscore.ConnectorBinding{}).
+		Watches(&walruscore.ConnectorBinding{}, &ctrlhandler.Funcs{
+			DeleteFunc: func(ctx context.Context, event ctrlevent.DeleteEvent,
+				limitingInterface workqueue.RateLimitingInterface,
+			) {
+				cb := event.Object.(*walruscore.ConnectorBinding)
+
+				ns := &core.Namespace{
+					ObjectMeta: meta.ObjectMeta{
+						Name: cb.Namespace,
+					},
+				}
+				err := r.client.Get(ctx, ctrlcli.ObjectKeyFromObject(ns), ns)
+				if err != nil {
+					ctrllog.FromContext(ctx).Error(err, "failed to get namespace", "namespace", cb.Namespace)
+				}
+
+				labels := ns.GetLabels()
+				delete(labels, walruscore.ProviderLabelPrefix+strings.ToLower(cb.Status.Type))
+				ns.SetLabels(labels)
+
+				err = r.client.Update(ctx, ns)
+				if err != nil {
+					ctrllog.FromContext(ctx).Error(err, "failed to update namespace", "namespace", cb.Namespace)
+				}
+			},
```

Comment:
Avoid doing heavy processing during an event transiting.

Use finalizer to remove/update the related resources.

Reply:
see walrus/subject.authz or project_subject.authz